### PR TITLE
pid1 serialization/deserialization fixes

### PR DESCRIPTION
**Pull request by poettering**

### CODE REVIEW

**Comment by evverx** [comment]
This pull request **introduces 1 alert** when merging 134dbf5e5ae9f42211869ce33378d721e361fe04 into 4e412d2684519b6054a47866196b88d58b2fa3cf - [view on LGTM.com](https://lgtm.com/projects/g/systemd/systemd/rev/pr-352ffb7d76909a4c7ad4e1fda2e266b67e609bcf)

**new alerts:**

* 1 for Comparison result is always the same

---

*Comment posted by [LGTM.com](https://lgtm.com)*

File: src/basic/exec-util.c [hunk 1]
```diff
@@ -294,27 +296,35 @@ static int gather_environment_collect(int fd, void *arg) {
...
+                if (r == 0)
+                        break;
+
+                assert_se(v = startswith(line, "env="));
 
-                k = deserialize_environment(env, line);
+                k = deserialize_environment(v, env);
                 if (k < 0)
-                        log_error_errno(k, "Invalid line \"%s\": %m", line);
+                        return k;
```
Comment (keszybz) [inline src/basic/exec-util.c]: LGTM is right, the condition is borked here.

**Comment by poettering** [comment]
I force-pushed a new version now, that hopefully fixes the LGTM issue. No changes besides this, i.e. not changes outside of exec-util.c

File: docs/CODING_STYLE [hunk 2]
```diff
@@ -458,3 +458,7 @@
 - When referring to a file system path that is a directory, please always
   suffix it with "/", to indicate that it is a directory, not a regular file
   (or other file system object).
+
+- Don't use fgets(), it's too hard to properly handle errors such as overly
+  long lines. Use read_line() instead, which is our own function that handles
+  this much nicer.
```
Comment (evverx) [inline docs/CODING_STYLE:464]: I've just opened https://github.com/systemd/systemd/pull/10533, which should help to enforce this rule more or less automatically.

**Comment by carnil** [comment]
[CVE-2018-15686](https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2018-15686) was assigned for this issue

**Comment by ChenQi1989** [comment]
@poettering
I'm doing backport to v239 to fix the CVE. I'd like to check with you whether I should backport all these 8 patches, or just the one that fixes the CVE.
In particular, the item 'rework serialization' worries me a little bit.

**Comment by ChenQi1989** [comment]
I realize I might not express it clearly enough.
For CVE-2018-15686 alone, instead of the serialization/deserialization mechanism in systemd, is the following patch enough?
core: when deserializing state always use read_line(…, LONG_LINE_MAX, …)

**Comment by yuwata** [comment]
@ChenQi1989 BTW, v239-stable branch has included this PR. https://github.com/systemd/systemd-stable/commits/v239-stable

**Comment by ChenQi1989** [comment]
This is the first time that I know there exists a systemd-stable repo... Thanks for the info.
I have a question.
For each released version, e.g. v239, how long will it be maintained?

**Comment by anarcat** [comment]
@yuwata which commits, in particular, fix this issue in 239? is https://github.com/systemd/systemd-stable/commit/1a05ff4948d778280ec155a9abe69d3360bfddd9 sufficient?

**Pull request merged**